refactor(agent): Modify agent status handling and remove tmp files

rene-oromtz opened the pull request:
## Description
The purpose of this PR is to improve the status handling for an agent. This is being done by removing the necessity of the creation/deletion of temporary files that can only be available by accessing the agent Charm Unit. 

The agent will now be able to get its state directly by querying the MongoDB by using server recent endpoint addition `GET /agents/data/<agent_name>` and process jobs depending on the state retrieved. 

Non breaking changes are introduced to perform the follows:
1. Whenever setting an agent state, the reason for the state is displayed in the UI. 
2. Agent is now able to detect if state is either `offline` or `maintenance` directly from the DB, in any case, the agent will stop looking for available jobs. 
3. Restart is also handled directly by temporarily setting a `restart` status on the agent. 
4. If at any point an offline/restart is detected, the new AgentStatusHandler will be able to carry over the reason for the offline/restart and only perform the action after the agent is not processing any job. 
<!--
Describe your changes here:

- What's the problem solved (briefly, since the issue is where this is elaborated in more detail).
- Introduce your implementation approach in a way that helps reviewing it well.
-->

## Resolved issues
Resolves [CERTTF-645](https://warthogs.atlassian.net/browse/CERTTF-645)
<!--
- Note the Jira and GitHub issue(s) resolved by this PR (`Fixes|Resolves ...`).
- Make sure that the linked issue titles & descriptions are also up to date.
-->

## Documentation
Documention needs to be updated to reflect how the agent is set to offline/restart/online, however, I'm making those modification once I completed the PR for the CLI changes. 
<!--
Please make sure that...
- Documentation impacted by the changes is up to date (becomes so, remains so).
  - Public documentation is in the repository in the docs/ folder.
  - If there impacts on Canonical processes the relevant documentation should be update outside the repository, confirm that this has happened.
- Tests are included for the changed functionality in this PR. If to be merged without tests, please elaborate why.
-->

## Web service API changes

<!--
- Are there new endpoints introduced? Please detail for each of them...
  - The rationale for introducing it
  - The interface
    - the HTTP verb
    - the URL structure
    - Versioning
      - Is it expected to be long time stable? It should be versioned (e.g. `.../v2/...`)
      - ... or it expected to evolve, perhaps expected to be a system internal need or something we are expecting to iterate on still? It should be marked unstable (e.g. `.../unstable/...`)
    - the possible request body
    - the response bodies and status code(s) for success and possible failure case(s)
  - Authorization
    - What credentials are required to access the resource?
    - What automated test scenarios are included for authorization failures?
- Are there changed database queries? (... which could cause performance regressions)
- Are there required configuration changes?
- Are there DB migrations included?
- ... or other things you'd like to be aware of at deploy time?
-->

## Tests
This was tested and validated in Staging

Agent Offline:
![Screenshot from 2025-06-17 14-44-28](https://github.com/user-attachments/assets/12a048df-5092-40ed-8c67-26e8c0b7bb68)

Attempting to submit job
```
testflinger-cli --server=https://testflinger-staging.canonical.com submit -p ~/Desktop/staging.yaml
ERROR: No online agents available for queue 202409-35503. If you want to wait for agents to become available, use the --wait-for-available-agents option.
```

If at any point the restart is detected, a note will be shown with the current phase state:
![Screenshot from 2025-06-17 15-02-14](https://github.com/user-attachments/assets/6c579a22-a247-47bd-a01c-c9be3f23409f)

Agent will continue processing the job and restart only when its safe to do so:
```
[25-06-17 20:48:50]    INFO: (agent.py:259)| Starting job 0268999f-01b0-4512-8736-b02a2312a8c0
[25-06-17 20:48:52]    INFO: (job.py:87)| Running setup_command: tf-setup
[25-06-17 20:49:07]    INFO: (job.py:87)| Running provision_command: tf-provision
[25-06-17 21:01:51]    INFO: (job.py:64)| No firmware_update_command configured, skipping...
[25-06-17 21:01:51]    INFO: (job.py:87)| Running test_command: tf-test
[25-06-17 21:04:34]    INFO: (job.py:74)| No allocate_data defined in job data, skipping...
[25-06-17 21:04:36]    INFO: (job.py:87)| Running reserve_command: tf-reserve
[25-06-17 21:05:33]    INFO: (job.py:87)| Running cleanup_command: tf-cleanup
[25-06-17 21:05:49]    INFO: (client.py:252)| Submitting job outcome for job: 0268999f-01b0-4512-8736-b02a2312a8c0
[25-06-17 21:05:50]    INFO: (agent.py:186)| Restarting agent
[25-06-17 21:05:53]   ERROR: (client.py:73)| InfluxDB host undefined
[25-06-17 21:05:53]    INFO: (__init__.py:138)| Checking jobs
```
<!--
- How was this PR tested? Please provide steps to follow so that the reviewer(s) can test on their end.
-->


[CERTTF-645]: https://warthogs.atlassian.net/browse/CERTTF-645?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

pedro-avalos commented on agent/src/testflinger_agent/agent.py:
```diff
@@ -124,66 +127,86 @@ def _post_initial_agent_data(self):
-        # Create the offline file, this should work even if it exists
-        open(self.get_offline_files()[0], "w").close()
+        # Offline set by server
+        if (
+            agent_state == AgentState.OFFLINE
+            or agent_state == AgentState.MAINTENANCE
+        ):
```
Might need formatting, but this would be easier to manage in case we ever add more "offline" states

```suggestion
        if agent_state in (
            AgentState.OFFLINE,
            AgentState.MAINTENANCE,
        ):
```

pedro-avalos commented on agent/src/testflinger_agent/agent.py:
```diff
@@ -124,66 +127,86 @@ def _post_initial_agent_data(self):
+        and requested by user.
+        """
+        logger.info("Restarting agent")
+        self.set_agent_state(AgentState.OFFLINE, comment)
```
Is this supposed to be `AgentState.RESTART`?

If not, could you add a short explanation as to why `OFFLINE` is used when marking the agent for restart?

pedro-avalos commented on agent/src/testflinger_agent/agent.py:
```diff
@@ -124,66 +127,86 @@ def _post_initial_agent_data(self):
+        and requested by user.
+        """
+        logger.info("Taking agent offline")
+        self.set_agent_state(AgentState.OFFLINE, comment)
+        # Need to set the offline flag to False to allow recovery
+        self.status_handler.update(offline=False, comment=comment)
```
Are the `offline_agent` and `restart_agent` functions flipped?

rene-oromtz replied:
Both are correct, restart will terminate agent code with sys.exit and offline will only stop processing jobs (it do needs to clear the flag from the handler, otherwise there is no way to recover the agent without manual intervention)

pedro-avalos commented on agent/src/testflinger_agent/agent.py:
```diff
@@ -311,7 +372,9 @@ def process_jobs(self):
                         # In this case, we need to mark the device offline
                         if exit_code == 46:
-                            self.mark_device_offline()
+                            comment = "Set to offline by agent. Recovery "
+                            f"Failed found during {job.job_id} execution."
```
I think this phrasing might be a little easier to read

```suggestion
                            comment = (
                                "Set to offline by agent. Recovery failed"
                                f" during job '{job.job_id}' execution."
                            )
```

pedro-avalos commented on agent/src/testflinger_agent/handlers.py:
```diff
@@ -31,3 +31,57 @@ def __init__(self, log_file: str):
+    def update(
+        self,
+        comment: str,
+        restart: bool | None = None,
+        offline: bool | None = None,
```
Since both of these can be `None` (which is considered the same as `False` in `if` statements by Python) perhaps you should have the following lines before the `if` statements:

```python
if restart is None and offline is None:
    raise ValueError("One of `restart` or `offline` required")
```

pedro-avalos commented on agent/src/testflinger_agent/handlers.py:
```diff
@@ -31,3 +31,57 @@ def __init__(self, log_file: str):
+            self.needs_offline = True
+            self.comment = comment
+        # Clear the flag if received an offline False
+        elif not offline and self.needs_offline:
```
Follow-up from the `None` comment I made earlier. If the caller does not specify `offline` (i.e. it's `None`), then should this line execute? Or, in other words, does `offline=None` / not providing `offline` imply that the agent should not be offline?

If that's the case, then I'm not quite sure why `restart` and `offline` are optional/`None`, maybe they should default to `False`, depending on the expected behavior

rene-oromtz replied:
Ah this is tricky. 
Initially I have them both as `False` however, there is an edge case where if during testing the agent is marked for offline and then, it also receives a restart it will only preserve the restart as the handler will clear the offline flag. The reason is this piece of code:
```
        # Clear the flag if received an offline False
        elif not offline and self._needs_offline:
            self._needs_offline = False
```

This is needed to clear the offline flag after effectively take the agent as offline. 

Another option I can do (if setting both to False is preferred) is that whenever calling the status handler for restart I should also be specifying the current offline flag. Something as:
```
self.status_handler.update(
    restart=needs_restart, 
    offline=self.status_handler.needs_offline, 
    comment=restart_comment
)
```

Let me try this scenario to see if it look cleaner

pedro-avalos commented on agent/src/testflinger_agent/handlers.py:
```diff
@@ -31,3 +31,57 @@ def __init__(self, log_file: str):
+        elif not offline and self.needs_offline:
+            self.needs_offline = False
+
+    def marked_for_restart(self) -> bool:
+        """Indicate the current restart state of the restart handler.
+
+        :return: True if a restart is neeeded, False otherwise.
+        """
+        return self.needs_restart
+
+    def marked_for_offline(self) -> bool:
+        """Indicate the current offline state of the offline handler.
+
+        :return: True if a offline is neeeded, False otherwise.
+        """
+        return self.needs_offline
+
+    def get_comment(self) -> str:
+        """Retrieve the comment from the status handler.
+
+        :return: Preserved comment if an agent status was modified.
+        """
+        return self.comment
```
I feel like these functions and "public" fields can be confusing. If what you want is only "getters" for `needs_restart` and `needs_offline`, I think you may want to use properties:

```python
class AgentStatusHandler:
    def __init__(self):
        self._needs_restart = False
        self._needs_offline = False
        self._comment = ""

    @property
    def needs_restart(self) -> bool:
        """Indicate the current restart state."""
        return self._needs_restart

    @property
    def needs_offline(self) -> bool:
        """Indicate the current offline state."""
        return self._needs_offline

    @property
    def comment(self) -> str:
        """Retrieve the comment on the state."""
        return self._comment
```

Another option is to allow using the instance variables `needs_offline`, `needs_restart`, and `comment` directly, but I do not think that's desirable, as updating them needs some extra logic (thus the `update` function)

rene-oromtz replied:
Thanks for the recommendation! I like this approach and implemented in the last commit

pedro-avalos reviewed (changes requested):
I have left a few comments regarding the code. There were 2 functions that either need comments or accidentally got flipped (`offline_agent` and `restart_agent`). I also left some questions regarding:

- one minor efficiency improvement
- style suggestions for parameters and class structure

rene-oromtz commented:
Thanks for the review! I added the suggestions and reformatted a bit for defaulting `offline` and `restart` to False. I tested on Staging and found couple of bugs which were also addressed. 
In case anything else is needed, please let me know!

pedro-avalos reviewed (dismissed):
LGTM

Thanks for the changes! I'm so glad to see the `/tmp` files go away

rene-oromtz commented:
@pedro-avalos I was about to merge this when I realize about a bug in a edge case. 
The scenario is the following:
1. Agent processing job. 
2. Received offline signal
3. Offline successfully deferred until job is no longer processing. 
4. Agent successfully set to offline
5. Agent successfully set back to online/waiting  from the CLI
6. On the next job the UI will display the preserved comment in the status handler as this wasn't cleared!

This is not affecting the agent as it will process the job and the comment will be cleared after completing this second job... But I found it too annoying as the comment is now displayed in the UI. 

Pushing the fix dismissed your review automatically, can you please help me with a re-review ?
